flycheck-elsa.el

```diff
@@ -46,28 +46,119 @@
+      (if (flycheck-elsa-cask-p)
+          (when-let (file (locate-dominating-file (buffer-file-name) "Cask"))
+            (file-name-directory file))
+        (file-name-directory (cdr (project-current))))
     default-directory))
 
+(defun flycheck-elsa--cask-command ()
+  "Return a list of strings to start Elsa within a Cask project."
+  '("cask" "exec" "elsa"))
+
+(defun flycheck-elsa--elsa-dependency-directory (library-name)
+  "Return the directory containing LIBRARY-NAME."
+  (file-name-directory (find-library-name library-name)))
+
+(defun flycheck-elsa--elsa-directory ()
+  "Return the directory where Elsa is installed."
+  (flycheck-elsa--elsa-dependency-directory "elsa"))
+
+(defun flycheck-elsa--elsa-dependencies ()
+  "Return a list of library names, as strings, that Elsa depend upon.
+The list does not contain Emacs itself."
+  (let* ((default-directory (flycheck-elsa--elsa-directory))
+         (package-description (with-temp-buffer
+                                (dired-mode) ;; required by `package-dir-info'
+                                (package-dir-info)))
+         (dependencies (mapcar (lambda (requirement) (symbol-name (car requirement)))
+                               (package-desc-reqs package-description))))
+    (seq-remove
+     (lambda (name) (string= name "emacs"))
+     dependencies)))
+
+(defun flycheck-elsa--elsa-dependency-directories ()
+  "Return a list of directories where Elsa and its dependencies are installed."
+  (let ((directories (mapcar
+                      #'flycheck-elsa--elsa-dependency-directory
+                      `("elsa" ,@(flycheck-elsa--elsa-dependencies)))))
+    (seq-uniq directories)))
+
+(defun flycheck-elsa--elsa-command ()
+  "Return the path to elsa executable file."
+  (expand-file-name
+   "elsa"
+   (expand-file-name "bin" (flycheck-elsa--elsa-directory))))
+
+(defun flycheck-elsa--host-command ()
+  "Return a list of strings to start Emacs with Elsa.
+Elsa and its dependencies are expected to be in `load-path'."
+  `(,(flycheck-elsa--elsa-command)
+    "--directory" "."
+    ,@(mapcan
+       (lambda (directory) (list "--directory" directory))
+       (flycheck-elsa--elsa-dependency-directories))))
+
+(defun flycheck-elsa-command ()
+  "Return a list of strings for the command to execute and its arguments.
+The result depends on the variable `flycheck-elsa-command'."
+  (cond
+   ((eq flycheck-elsa-command 'cask) (flycheck-elsa--cask-command))
+   ((eq flycheck-elsa-command 'host) (flycheck-elsa--host-command))
+   (t flycheck-elsa-command)))
+
+(defun flycheck-elsa-cask-p ()
+  "Return non-nil if the project should be checked through Cask."
+  (eq flycheck-elsa-command 'cask))
+
 (flycheck-define-checker emacs-lisp-elsa
   "An Emacs Lisp checker using Elsa"
-  :command ("cask" "exec" "elsa" source)
+  :command (;; flycheck forces us to pass a string as first argument:
+            ;; https://github.com/flycheck/flycheck/issues/1515
+            "/usr/bin/env"
```

Comment:
Is this safe on every platform? Or do we completely ignore the executable here anyway so it doesn't really matter?

Reply:
The command is never executed so it is safe. The file should exist for the checker to be activated though. I think `/usr/bin/env` won't work on Windows because of that.

I replaced `/usr/bin/env` with `emacs` just like [what flycheck itself does](https://github.com/flycheck/flycheck/blob/master/flycheck.el#L7223).